[DYN-2680] Write logs to std out instead of creating new files for each test.

reddyashish opened the pull request:
### Purpose

This PR is in reference to the task: https://jira.autodesk.com/browse/DYN-2680. 

In the current design, a log file is created everytime dynamo is started and closed. So during a test run on the VM, around 5000-10000 logs are created. Over time, this is causing performance issues on the VM. We are not using these logs anyways so it is better to avoid it. 

After this change, during tests, we write the logs only to std out instead of file or to dynamo console. 

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo

mmisol commented on src/DynamoCore/Logging/DynamoLogger.cs:
```diff
@@ -157,7 +157,7 @@ public IEnumerable<NotificationMessage> StartupNotifications
         /// <param name="debugSettings">Debug settings</param>
         /// <param name="logDirectory">Directory path where log file will be written</param>
-        public DynamoLogger(DebugSettings debugSettings, string logDirectory)
+        public DynamoLogger(DebugSettings debugSettings, string logDirectory, Boolean IsTestMode = false)
```
Should we used chained constructors instead? Optional parameters require a client recompile, so technically this is a breaking change.

mmisol replied:
NIT: The parameter name should start with lower case: `isTestMode`

mjkkirschner replied:
I finally found an official version of this:
https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/breaking-change-rules.md

but It does not seem as comprehensive as:
https://stackoverflow.com/questions/1456785/a-definitive-guide-to-api-breaking-changes-in-net

please review it whenever making changes to public apis.
relevant section:
`Adding a parameter with a default value.`
though it requires careful reading.

reddyashish replied:
Yes will be changing this to an overloaded constructor. I tried to look if adding a optional parameter is a good design or not for public API's but that link is really comprehensive

mmisol commented on src/DynamoCore/Logging/DynamoLogger.cs:
```diff
@@ -197,6 +202,13 @@ private void Log(string message, LogLevel level, bool reportModification)
+                {
+                    Console.WriteLine(string.Format("{0} : {1}", DateTime.UtcNow.ToString("u"), message));
+                    return;
+                }
```
Why not just leverage on the existing code for the console logging by setting the `level` variable to `LogLevel.Console` if `testMode` is true?

reddyashish replied:
The LogLevel.Console is configured to write to both the dynamo console and to the log file. So could't use it.

mmisol replied:
You are indeed right. The fact that it does both things having that name is confusing :)

reddyashish commented:
Fixed a couple of existing tests. These tests verify that the logs are written to dynamo console and to a file. When we changed the behaviour to write the logs to std out in test mode, they were failing. Fixed them accordingly.

mmisol commented on test/DynamoCoreTests/Logging/DynamoLoggerTest.cs:
```diff
@@ -45,27 +63,9 @@ public void test_dynamologger_LogEventArgs()
-                CorePath = config.DynamoCorePath,
-                HostPath = config.DynamoHostPath,
-                PathResolver = config.PathResolver
-            });
```
Just to confirm, these params were not needed, right?

reddyashish replied:
yeah, these were null values.

mmisol commented on test/DynamoCoreTests/Migration/WorkspaceMigrationsTests.cs:
```diff
@@ -15,21 +16,31 @@ class WorkspaceMigrationsTests : DynamoModelTestBase
-            Assert.IsTrue(File.Exists(currentLoggerPath));//Check that the log was created successfully
-            Assert.IsTrue(CurrentDynamoModel.Logger.LogText.Contains("migration from 0.5.3.x to 0.6.1.x"));//Checks that the log text contains the migration info
+            //Checks that the console output text contains the migration info
+            Assert.IsTrue(consoleOutput.Contains("migration from 0.5.3.x to 0.6.1.x"));
```
Maybe for this test we actually want to use the non-test version of the logger? I'm thinking that having a persistent log after a migration is done is desired and should be tested to exist.

reddyashish replied:
The thing is, the dynamo logger is created from the dynamo model when the dynamo model is initialized from the tests. It is a read only property and we create a instance of logger in non test mode here and add it to the model

mjkkirschner commented on test/DynamoCoreTests/Migration/WorkspaceMigrationsTests.cs:
```diff
@@ -15,21 +16,31 @@ class WorkspaceMigrationsTests : DynamoModelTestBase
+
+                consoleOutput = stringWriter.ToString();
+
+                StreamWriter sw = new StreamWriter(Console.OpenStandardOutput());
```
should you call dispose on StreamWriter? I believe it implements IDisposable!

mjkkirschner replied:
sorry, looked too quickly, didn't see this was one test, not the logger!

reddyashish replied:
We can avoid creating the StreamWriter object. Saving the std out beforehand and setting it back.
